raft/fsm.cc

```diff
@@ -205,10 +205,7 @@ void fsm::become_candidate(bool is_prevote, bool is_leadership_transfer) {
     // and initiating another round of RequestVote RPCs.
     _last_election_time = _clock.now();
 
-    auto& votes = candidate_state().votes;
-
-    const auto& voters = votes.voters();
-    if (!voters.contains(server_address{_my_id})) {
+    if (!_log.get_configuration().can_vote(_my_id)) {
```

Comment:
You basically make this if always false for the case when it matters. You should check configuration().current, not configuration joint.

Reply:
Basically, if you read the comment, it says it all. If the current configuration doesn't contain self, but hte previous does, *and* joint is not committed yet, we should still be a leader. 

Your change is not quite the same. E.g. you will become a candidate if you couldn't be a voter in neither C_new nor C_old, but the current configuration transition is not committed yet.

Reply:
@kbr-  Why do you need this change at all? Both codes looks identical to me.

Reply:
> You basically make this if always false for the case when it matters. You should check configuration().current, not configuration joint.

Why? If a node is a voter in an active config it should not enter the `if`.

Reply:
Indeed. So the check should be that a node is not a voter in the current config but was a voter in the previous config. This node should enter the if.

Reply:
> @kbr- Why do you need this change at all? Both codes looks identical to me.

True, it's not needed, it's a leftover from my earlier experimentation.

Reply:
> Your change is not quite the same. E.g. you will become a candidate if you couldn't be a voter in neither C_new nor C_old, but the current configuration transition is not committed yet.

The second if will prevent me from becoming a candidate in the case you describe (I wasn't a voter in C_old, so I don't become a candidate):
```cpp
        auto prev_cfg = _log.get_prev_configuration();
        assert(prev_cfg);
        if (!prev_cfg->can_vote(_my_id)) {
            // We weren't a voter before.
            become_follower(server_id{});
            return;
}
```

Also as Gleb observed, this change is not a change at all (given the new definition of `voters()`)

Reply:
> Indeed. So the check should be that a node is not a voter in the current config but was a voter in the previous config. This node should enter the if.

Why? If a node can become a voter it should not enter this `if`.

Reply:
This is why here: https://github.com/scylladb/scylla/pull/10640#discussion_r880182879 I wrote "active" :)